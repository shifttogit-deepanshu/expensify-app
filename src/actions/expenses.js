import database from "../firebase/firebase"
export const addExpense = (expense)=>(
    {   type:"ADD_EXPENSE",
        expense
    }

)

export const startAddExpense = (expenseData = {})=>{
    return (dispatch)=>{
        const {
            description='',
            note='',
            amount=0,
            createdAt=0
        }=expenseData
        const expense = {description,note,amount,createdAt};
    return database.ref('expenses').push(expense).then((ref)=>{
        dispatch(addExpense({
            id:ref.key,
            ...expense
        }))
    })
    }
}

export const removeExpense = ({id=0}={})=>(
    {
        type:"REMOVE_EXPENSE",
        id
    }
)

export const editExpense = (id,updates)=>({
    type:"EDIT_EXPENSE",
    id,
    updates
})

export const setExpense = (expenses)=>({
    type:"SET_EXPENSE",
    expenses
})

export const startSetExpense=()=>{
    return (dispatch)=>{
        return database.ref('expenses').once('value').then((snapshot)=>{
            const expenses = []
            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                   id: childSnapshot.key,
                   ...childSnapshot.val()
                })
            })
            dispatch(setExpense(expenses))
        })
    }
}