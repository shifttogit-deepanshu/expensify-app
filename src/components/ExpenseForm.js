import React from "react"

import { SingleDatePicker} from 'react-dates';

import moment from "moment"

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)
      
        this.state={
            description:props.expense ? props.expense.description: '',
            notes:props.expense ? props.expense.notes : '',
            amount:props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt:props.expense ? moment(props.expense.createdAt) : moment(),
            focused:false,
            error:''
        }

    }
   
    onDescriptionChange = (e)=>{
        const description = e.target.value
        this.setState(()=>({description}))
    }
    onNotesChange = (e)=>{
        const notes = e.target.value
        this.setState(()=>({notes}))
    }
    onAmountChange = (e)=>{
        const amount = e.target.value
        if(!amount || amount.match(/^\d+(\.\d{0,2})?$/)){
        this.setState(()=>({amount}))
        }
    }
    onDateChange = (createdAt)=>{
        if(createdAt){
        this.setState(()=>({createdAt}))
        }
    }
    onFocusChange = ({focused})=>{
        this.setState(()=>({focused}))
    }
    onSubmit = (e)=>{
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:"Description and amount cannot be left blank"}))
        }
        else{
            console.log("submitted")
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                notes:this.state.notes
            
            })
        }
    }
    render(){
        return (
            <div>
            <h2>{this.state.error}</h2>
            <form onSubmit={this.onSubmit}>

            <input type="text" placeholder="description" autoFocus value={this.state.description} 
            onChange={this.onDescriptionChange}/><br/>
            <input type="number" placeholder="amount" value={this.state.amount} onChange={this.onAmountChange}/><br/>
            <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
            /><br/>
            <textarea placeholder="add note here" value={this.state.notes} onChange={this.onNotesChange}></textarea><br/>
            <button>Add Expense</button>
            </form>

            </div>
        )
    }
}