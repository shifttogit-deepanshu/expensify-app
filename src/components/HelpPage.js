import React from 'react'

const HelpPage = ()=>{
 
    return <h2>This is HelpPage</h2>
}

export default HelpPage;