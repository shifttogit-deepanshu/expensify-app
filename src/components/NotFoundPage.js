import React from 'react'
import {Link} from "react-router-dom";

const NotFoundPage = ()=>{
    return <div>
    <h2>404! -</h2> <Link to="/">Go TO HOME</Link>
    </div>
}

export default NotFoundPage;