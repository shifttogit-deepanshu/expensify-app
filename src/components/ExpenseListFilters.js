import React from "react"
import {connect} from "react-redux"
import {setTextFilter} from "../actions/filter"
import {sortByDate} from "../actions/filter"
import {sortByAmount} from "../actions/filter"
import {DateRangePicker} from "react-dates"
import {setStartDate} from "../actions/filter"
import {setEndDate} from "../actions/filter"

export class ExpenseListFilters extends React.Component{
    constructor(props){
        super(props)
        this.state={
            focusedInput:null
        }
    }
    onDatesChange = ({startDate,endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onTextChange=(e)=>{
        this.props.setTextFilter(e.target.value)
    }
    onDateFilterChange=(e)=>{
        if(e.target.value=="date"){
            this.props.sortByDate()
        }
        else if(e.target.value=="amount"){
            this.props.sortByAmount()
        }}
    render(){
    return (<div>
    <input type="text" value={this.props.filter.text} onChange={this.onTextChange} />
    <select value={this.props.filter.sortBy} onChange={this.onDateFilterChange}>
    <option value="date">Date</option>
    <option value="amount">Amount</option>
    </select>
    <DateRangePicker
    startDate={this.props.filter.startDate}
    startDateId='startTheDate'
    endDate={this.props.filter.endDate}
    endDateId="endTheDate"
    onDatesChange={this.onDatesChange}
    focusedInput={this.state.focusedInput}
    onFocusChange={focusedInput => this.setState({ focusedInput })}
    showClearDates={true}
    isOutsideRange={()=>false}

    numberOfMonths={1}

    />
    </div>)
    }
}

const StateToProps = (state)=>{
    return {
        filter:state.filter
    }
}
const mapDispatchtoProps = (dispatch)=>{
    return {
        setStartDate: (startDate)=>{dispatch(setStartDate(startDate))},
        setEndDate:(endDate)=>{dispatch(setEndDate(endDate))},
        setTextFilter:(data)=>{dispatch(setTextFilter(data))},
        sortByDate:()=>{dispatch(sortByDate())},
        sortByAmount:()=>{dispatch(sortByAmount())}

    }
}

export default connect(StateToProps,mapDispatchtoProps)(ExpenseListFilters)

