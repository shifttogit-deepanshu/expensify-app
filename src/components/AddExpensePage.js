import React from 'react'
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {addExpense} from "../actions/expenses"


const AddExpensePage = (props)=>{
    
    return <div>
    <h2>This is AddExpensePage</h2>
    <ExpenseForm onSubmit={(expense)=>{
        props.dispatch(addExpense(expense))
        props.history.push('/')
    }}/>
    </div>
    
}

export default connect()(AddExpensePage);