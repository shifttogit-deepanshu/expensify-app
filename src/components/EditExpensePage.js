import React from 'react'
import ExpenseForm from "./ExpenseForm"
import {connect}  from "react-redux"
import {editExpense} from "../actions/expenses"
import { removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component{
    onSubmit=(expense)=>{
        this.props.editExpense(this.props.expense.id,expense)
        this.props.history.push('/')

    }
    buttonClicked=() => {
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/')
      }
    

    render() {return (<div>
    <h2>This is EditExpensePage</h2>
    <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
    <button onClick={this.buttonClicked}>Remove</button>
    </div>)
    }
}
const mapStateToProps = (state,props)=>{
    return {
        expense:state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}
const mapDispatchToProps = (dispatch)=>{
    return {
        editExpense:(id,expense)=>dispatch(editExpense(id,expense)),
        removeExpense:(id)=>dispatch(removeExpense({id}))
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)