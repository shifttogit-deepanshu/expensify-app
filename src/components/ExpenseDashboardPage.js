import React from 'react'
import ExpenseList from "./ExpenseList"
import ExpenseListFilters from "./ExpenseListFilters"

const ExpenseDashboardPage = ()=>{
    return <div>
    <h2>This is ExpenseDashboardPage</h2>
    <ExpenseListFilters />
    <ExpenseList />
    </div> 
}

export default ExpenseDashboardPage;