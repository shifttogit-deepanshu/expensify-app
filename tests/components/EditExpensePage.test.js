import React from 'react'
import {shallow} from 'enzyme'
import {EditExpensePage} from '../../src/components/EditExpensePage'
import expenses from '../fixtures/expenses'

let editExpense,history,removeExpense,wrapper

beforeEach(()=>{
    editExpense = jest.fn()
    history = {push:jest.fn()}
    removeExpense = jest.fn()
    wrapper = shallow(<EditExpensePage 
        editExpense={editExpense} 
        history={history} 
        removeExpense={removeExpense}
        expense={expenses[1]} />)

})

test('should render page correctly',()=>{
    
    expect(wrapper).toMatchSnapshot()
})

test('should run editExpense function correctly',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id,expenses[1])
    expect(history.push).toHaveBeenLastCalledWith('/')
})

test('should remove expense correcty',()=>{
    wrapper.find('button').simulate('click')
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[1].id)
})

