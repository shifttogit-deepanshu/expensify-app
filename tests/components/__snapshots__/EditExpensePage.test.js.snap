// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render page correctly 1`] = `
<div>
  <h2>
    This is EditExpensePage
  </h2>
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": 0,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
